lib/Conversion/HWToBTOR2/HWToBTOR2.cpp

```diff
@@ -0,0 +1,817 @@
+    }
+
+    // Add the width to the declared sorts along with the associated line id
+    sortToLIDMap[width] = lid;
+
+    // Build and return a sort declaration
+    os << lid++ << wsStr << sortStr << wsStr << type << wsStr << width << nlStr;
+  }
+
+  // Generates an input declaration given a sort lid and a name
+  void genInput(size_t width, StringRef name) {
+    // Retrieve the lid associated with the sort (sid)
+    size_t sid = sortToLIDMap.at(width);
+
+    // Generate input declaration
+    os << lid++ << wsStr << inputStr << wsStr << sid << wsStr << name << nlStr;
+  }
+
+  // Generates a constant declaration given a value, a width and a name
+  void genConst(int64_t value, size_t width, Operation *op) {
+    // For now we're going to assume that the name isn't taken, given that hw is
+    // already in SSA form
+    getOrCreateOpLID(op);
+
+    // Retrieve the lid associated with the sort (sid)
+    size_t sid = sortToLIDMap.at(width);
+
+    os << lid++ << wsStr << constdStr << wsStr << sid << wsStr << value
+       << nlStr;
+  }
+
+  // Generates a zero constant expression
+  void genZero(size_t width) {
+    // Check if the constant has been created yet
+    if (getConstLID(0, width) != noLID)
+      return;
+
+    // Retrieve the lid associated with the sort (sid)
+    size_t sid = sortToLIDMap.at(width);
+
+    // Keep track of this value in a constant declaration tracker
+    constToLIDMap[{0, width}] = lid;
+
+    // Build and return the zero btor instruction
+    os << lid++ << wsStr << zeroStr << wsStr << sid << nlStr;
+  }
+
+  // Generates a binary operation instruction given an op name, two operands and
+  // a result width
+  void genBinOp(StringRef inst, Operation *binop, Value op1, Value op2,
+                size_t width) {
+    // Set the LID for this operation
+    getOrCreateOpLID(binop);
+
+    // Find the sort's lid
+    size_t sid = sortToLIDMap.at(width);
+
+    // Assuming that the operands were already emitted
+    // Find the LIDs associated to the operands
+    size_t op1LID = getOrCreateOpLID(op1);
+    size_t op2LID = getOrCreateOpLID(op2);
+
+    // Build and return the string
+    os << lid++ << wsStr << inst << wsStr << sid << wsStr << op1LID << wsStr
+       << op2LID << nlStr;
+  }
+
+  // Generates a slice instruction given an operand, the lowbit, and the width
+  void genSlice(Operation *srcop, Value op0, size_t lowbit, int64_t width) {
+    // Assign a LID to this operation
+    getOrCreateOpLID(srcop);
+
+    // Find the sort's associated lid in order to use it in the instruction
+    size_t sid = sortToLIDMap.at(width);
+
+    // Assuming that the operand has already been emitted
+    // Find the LID associated to the operand
+    size_t op0LID = getOrCreateOpLID(op0);
+
+    // Build and return the slice instruction
+    os << lid++ << wsStr << sliceStr << wsStr << sid << wsStr << op0LID << wsStr
+       << (width - 1) << wsStr << lowbit << nlStr;
```

Comment:
This feels like it could subtly break: `getOrCreateOpLID` reads `lid` to associate it with the given op. The initial `getOrCreateOpLID(srcop);` associates `srcop` with the current `lid`, but doesn't really increment the `lid` in any way. Later on, `getOrCreateOpLID(op0);` would override the initial association between `(srcop,lid)` with `(op0,lid)`. The `lid` is only incremented during `os << ...` at the end here.

The comment mentions an assumption that the operand has already been emitted. Can you check that invariant in an assert somehow? If the operand hasn't been emitted, this would still produce output that looks superficially valid, and might be hard to debug.

Alternatively, could there be a `getOpLID(...)` function which asserts that the operand has already been emitted? If it hasn't, you could emit a `op->emitOpError("has not been emitted yet")` in that function and get nice runtime errors in case something breaks. Unless your pass guarantees that all operands have already been emitted -- then an assert is fine 🙂